text/0722-align-mount-route-apis.md

```diff
@@ -0,0 +1,73 @@
+---
+Stage: Accepted
+Start Date: 2021-02-12
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/722
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Align mount and route APIs
+
+## Summary
+
+Update the Ember `mount` API to accept a function argument which defines the Engine’s routes similar to how the `route` API works today.
+
+## Motivation
+
+Some Ember applications utilize in-repo ember-engines strictly for their lazy loading and code organization features and not for their isolation features. In this context the isolation actually causes some maintenance overhead and prevents build optimization.
+
+If isolation is not a requirement and all engines in an application are in-repo it makes more sense to define all of the routes at the application level in the router.js.
+
+## Detailed design
+
+The required change will be made to the `mount` method in [packages/@ember/-internals/routing/lib/system/dsl.ts ](https://github.com/emberjs/ember.js/blob/b35106e4d8471e396eb1ee5a5044b6bbe72fa069/packages/%40ember/-internals/routing/lib/system/dsl.ts#L176). A third argument will be added matching the method signature of the `route` method.
+
+```js
+mount(name, options, mountCallback) {...}
+```
+
+If the argument is present and it is a function the route-map will not be resolved and the callback will be used in place of the resolved route-map.
+
+```js
+let engineRouteMap = null;
+
+if (!mountCallback) {
+  engineRouteMap = this.options.resolveRouteMap(name);
+}
+
+if (mountCallback) {
+  mountCallback.call(childDSL);
+} else if (engineRouteMap) {
+  engineRouteMap.class.call(childDSL);
+}
+```
+
+## How we teach this
+
+The alignment of the `route` and `mount` APIs should be easy to teach and is likely already inline with peoples mental model of the `route` API.
+
+This change also unlocks a more idomatic way of using in-repo engines when the use-case does not require isolation.
```

Comment:
What is the impact here to `standalone engines`?

Reply:
Ya, we should make sure to mention it. Overall the impact to stand alone engines is essentially none. We _could_ suggest folks start leveraging the new callback to ensure the bundles are built properly (instead of "magically" hoisting `engine-name-here/routes` module into the host application).

Reply:
Looking better at this RFC, I'm not sure. Declaring the callback with child routes from engines seems to force the host app to know things about the engine (internal implementation about how I'm structuring my engine). We must maintain the principle of isolation here. 

We shouldn't confuse the `hosting` from **mountPoint** with the code isolation when the host app will know the implementation of the route from the engine and will declare it on `router.js`.

Reply:
> Looking better at this RFC, I'm not sure. Defines an engine's routing map from engines into router.js seems to force the host app to know things about the engine (internal implementation about how I'm structuring my engine). We must maintain the principle of isolation here.

I don't think I agree with you here. A non-trivial number of ember-engines users (most of the applications that use ember-engines that I work on in fact) only include in-repo engines where there is effectively no benefit to this type of isolation. Additionally, there is no runtime isolation from the engine's routes (all routes are already **required** to be included up front in the applications own asset bundle).

Reply:
> only include in-repo engines where there is effectively no benefit to this type of isolation

I think it's still a kind of isolation where people aren't forced to separate your code using standalone engines. If some user wants to break your app into 2 engines and keep it as in-repo-engines I don't think it's bad and we shouldn't expose the engine routing map only because it is inside the `/lib` folder. 

let me know if I'm missing something here @rwjblue you have been working on engines since the beginning, I may be missing something.